ipa-core/src/protocol/ipa_prf/oprf_padding/mod.rs

```diff
@@ -1,5 +1,814 @@
+        &padding_params,
+    )
+    .await?;
+
+    let after_padding_len = input.len();
+    tracing::info!(
+        "Total number of padding records added: {}. Padding Parameters: {:?}",
+        after_padding_len - initial_len,
+        padding_params
+    );
+
+    Ok(input)
+}
+
+/// Apply dp padding with one pair of helpers generating the noise
+/// Steps
+///     1.  Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+///         and use it to sample the same random noise for padding from `OPRFPaddingDp`.
+///         They will generate secret shares of these fake rows.
+///     2.  `h_i` and `h_i_plus_one` will send the send `total_number_of_fake_rows` to `h_out`
+///     3.  `h_out` will generate secret shares of zero for as many rows as the `total_number_of_fake_rows`
+///
+/// # Errors
+/// Will propogate errors from `OPRFPaddingDp`
+/// # Panics
+/// Will panic if called with Roles which are not all unique
+pub async fn apply_dp_padding_pass<C, BK, TV, TS, const B: usize>(
+    ctx: C,
+    mut input: Vec<OPRFIPAInputRow<BK, TV, TS>>,
+    h_i: Role,
+    h_i_plus_one: Role,
+    h_out: Role,
+    padding_params: &PaddingParameters,
+) -> Result<Vec<OPRFIPAInputRow<BK, TV, TS>>, Error>
+where
+    C: Context,
+    BK: BooleanArray + U128Conversions,
+    TV: BooleanArray,
+    TS: BooleanArray,
+{
+    // assert roles are all unique
+    assert!(h_i != h_i_plus_one);
+    assert!(h_i != h_out);
+    assert!(h_out != h_i_plus_one);
+
+    let mut total_number_of_fake_rows = 0;
+    let mut padding_input_rows: Vec<OPRFIPAInputRow<BK, TV, TS>> = Vec::new();
+
+    // Step 1: Helpers `h_i` and `h_i_plus_one` will get the same rng from PRSS
+    // and use it to sample the same random noise for padding from OPRFPaddingDp.
+    // They will generate secret shares of these fake rows.
+    if ctx.role() != h_out {
+        total_number_of_fake_rows = two_parties_add_dummies::<C, BK, TV, TS, B>(
+            &ctx,
+            &mut padding_input_rows,
+            h_i,
+            h_i_plus_one,
+            padding_params,
+        )?;
+    }
+
+    // Step 2: h_i and h_i_plus_one will send the send total_number_of_fake_rows to h_out
+    let send_ctx = ctx
+        .narrow(&SendTotalRows::SendFakeNumRecords)
+        .set_total_records(TotalRecords::ONE);
+    if ctx.role() == h_i {
+        let send_channel = send_ctx.send_channel::<BA32>(send_ctx.role().peer(Direction::Left));
+        let _ = send_channel
+            .send(
+                RecordId::FIRST,
+                BA32::truncate_from(u128::from(total_number_of_fake_rows)),
+            )
+            .await;
+    }
+    if ctx.role() == h_i_plus_one {
+        let send_channel = send_ctx.send_channel::<BA32>(send_ctx.role().peer(Direction::Right));
+        let _ = send_channel
+            .send(
+                RecordId::FIRST,
+                BA32::truncate_from(u128::from(total_number_of_fake_rows)),
+            )
+            .await;
+    }
```

Comment:
Can you combine these into a single code block? This is very repetitive. And maybe add a comment about the rationale for this (redundancy in case of a malicious helper)

Reply:
Okay, refactored a bit to remove some duplication.  Also added comment about why we check the value received from both other parties. 
```
    match ctx.role() {
        role if role == h_i || role == h_i_plus_one => {
            let direction = if role == h_i {
                Direction::Left
            } else {
                Direction::Right
            };
            let send_channel = send_ctx.send_channel::<BA32>(send_ctx.role().peer(direction));
            let _ = send_channel
                .send(
                    RecordId::FIRST,
                    BA32::truncate_from(u128::from(total_number_of_fake_rows)),
                )
                .await;
        }
        _h_out => {
            let recv_channel_right =
                send_ctx.recv_channel::<BA32>(send_ctx.role().peer(Direction::Right));
            let from_right = match recv_channel_right.receive(RecordId::FIRST).await {
                Ok(v) => u32::try_from(v.as_u128()).unwrap(),
                Err(e) => return Err(e.into()),
            };
            let recv_channel_left =
                send_ctx.recv_channel::<BA32>(send_ctx.role().peer(Direction::Left));
            let from_left = match recv_channel_left.receive(RecordId::FIRST).await {
                Ok(v) => u32::try_from(v.as_u128()).unwrap(),
                Err(e) => return Err(e.into()),
            };
            assert_eq!(from_right, from_left);
            total_number_of_fake_rows = from_right;
        }
    }
    ```